include/swift/AST/Attr.h

```diff
@@ -1711,6 +1712,150 @@ class DeclAttributes {
   SourceLoc getStartLoc(bool forModifiers = false) const;
 };
 
+/// A declaration name with location.
+struct DeclNameWithLoc {
+  DeclName Name;
+  DeclNameLoc Loc;
+};
+
+/// Attribute that marks a function as differentiable and optionally specifies
+/// custom associated derivative functions: 'jvp' and 'vjp'.
+///
+/// Examples:
+///   @differentiable(jvp: jvpFoo where T : FloatingPoint)
+///   @differentiable(wrt: (self, x, y), jvp: jvpFoo)
+class DifferentiableAttr final
+    : public DeclAttribute,
+      private llvm::TrailingObjects<DifferentiableAttr,
+                                    ParsedAutoDiffParameter> {
+  friend TrailingObjects;
+
+  /// Whether this function is linear (optional).
+  bool linear;
+  /// The number of parsed parameters specified in 'wrt:'.
+  unsigned NumParsedParameters = 0;
+  /// The JVP function.
+  Optional<DeclNameWithLoc> JVP;
+  /// The VJP function.
+  Optional<DeclNameWithLoc> VJP;
+  /// The JVP function (optional), resolved by the type checker if JVP name is
+  /// specified.
+  FuncDecl *JVPFunction = nullptr;
+  /// The VJP function (optional), resolved by the type checker if VJP name is
+  /// specified.
+  FuncDecl *VJPFunction = nullptr;
+  /// The differentiation parameters' indices, resolved by the type checker.
+  AutoDiffParameterIndices *ParameterIndices = nullptr;
```

Comment:
Note: `AutoDiffParameterIndices` has been removed (fully replaced with `AutoDiffIndexSubset`) on `tensorflow` branch as of https://github.com/apple/swift/pull/27514, in the spirit of reducing the number of upstreamed data structures.

Reply:
Thanks, Dan. I will go ahead and clean up this PR as well.

Reply:
This is done. I think it makes sense to upstream `AutoDiffIndexSubset` first. I am going split that off now.

Reply:
Upstreaming `AutoDiffIndexSubset` alone makes sense. Do you plan to work on it? If not, I can do that today.

Reply:
`AutoDiffIndexSubset` is being upstreamed in PR #27555.

Reply:
Thanks, @rxwei I will rebase this PR on top of it.